autosklearn/util/data.py

```diff
@@ -70,3 +132,223 @@ def predict_RAM_usage(X: np.ndarray, categorical: List[bool]) -> float:
             estimated_columns += 1
     estimated_ram = estimated_columns * X.shape[0] * X.dtype.itemsize
     return estimated_ram
+
+
+def subsample(
+    X: SUPPORTED_FEAT_TYPES,
+    y: Union[List, np.ndarray, pd.DataFrame, pd.Series],
+    is_classification: bool,
+    sample_size: Union[float, int],
+    random_state: Optional[Union[int, np.random.RandomState]] = None,
+) -> Tuple[SUPPORTED_FEAT_TYPES, Union[List, np.ndarray, pd.DataFrame, pd.Series]]:
+    """ Subsamples data returning the same type as it recieved.
+
+    If `is_classification`, we split using a stratified shuffle split which
+    preserves unique labels in the training set.
+
+    NOTE:
+    It's highly unadvisable to use lists here. In order to preserver types,
+    we convert to a numpy array and then back to a list.
+
+    NOTE2:
+    Interestingly enough, StratifiedShuffleSplut and descendants don't support
+    sparse `y` in `split(): _check_array` call. Hence, neither do we.
+
+    NOTE3:
+    The core autosklearn library doesn't rely on the full type of X.
```

Comment:
See early comment with regards to wide vs narrow typing of functions, while we don't use all the types it can accept in the library, this `subsample` function, as a standalone function, works with all types in it's signature.

Reply:
It might be good to narrow the typing here, though, as these notes are introduced by this PR.

Reply:
Settled on Narrow Typing, I will update the types to reflect this